docs/ql-design-patterns.md

```diff
@@ -0,0 +1,82 @@
+
+To allow both extensibility and refinement of classes, we use what is commonly referred to as the `::Range` pattern (since https://github.com/github/codeql/pull/727), but the actual implementation can use different names.
+
+<details>
+<summary>Generic example of how to define classes with ::Range</summary>
+
+Instead of
+```ql
+/** <QLDoc...> */
+abstract class MySpecialExpr extends Expr {
+  /** <QLDoc...> */
+  abstract int memberPredicate();
+}
+```
+with
+```ql
+class ConcreteSubclass extends MySpecialExpr { ... }
+```
+
+use
+
+```ql
+/**
+ * <QLDoc...>
+ *
+ * Extend this class to refine existing API models. If you want to model new APIs,
+ * extend `MySpecialExpr::Range` instead.
+ */
+class MySpecialExpr extends Expr {
+  MySpecialExpr::Range range;
+
+  MySpecialExpr() { this = range }
+
+  /** <QLDoc...> */
+  int memberPredicate() { result = range.memberPredicate() }
+}
+
+/** Provides a class for modeling new <...> APIs. */
+module MySpecialExpr {
+  /**
+   * <QLDoc...>
+   *
+   * Extend this class to model new APIs. If you want to refine existing API models,
+   * extend `MySpecialExpr` instead.
+   */
+  abstract class Range extends Expr {
+    /** <QLDoc...> */
+    abstract int memberPredicate();
+  }
+}
+```
+with
+```ql
+class ConcreteSubclass extends MySpecialExpr::Range { ... }
+```
+
+</details>
+
+Note, previously we used to write `MySpecialExpr() { this = self }`, but we now recommend using `MySpecialExpr() { this = range }` instead to avoid anyone mistakenly thinking that `self` and `this` are synonyms in general.
+
+### Rationale
+
+Let's use an example from the Go libraries: https://github.com/github/codeql-go/blob/2ba9bbfd8ba1818b5ee9f6009c86a605189c9ef3/ql/src/semmle/go/Concepts.qll#L119-L157
+
+`EscapeFunction`, as the name suggests, models various APIs that escape meta-characters. It has a member-predicate `kind()` that tells you what sort of escaping the modelled function does. For example, if the result of that predicate is `"js"`, then this means that the escaping function is meant to make things safe to embed inside JavaScript.
+`EscapeFunction::Range` is subclassed to model various APIs, and `kind()` is implemented accordingly.
+But we can also subclass `EscapeFunction` to, as in the above example, talk about all JS-escaping functions.
+
+You can, of course, do the same without the `::Range` pattern, but it's a little cumbersome:
+If you only had an `abstract class EscapeFunction { ... }`, then `JsEscapeFunction` would need to be implemented in a slightly tricky way to prevent it from extending `EscapeFunction` (instead of refining it). You would have to give it a charpred `this instanceof EscapeFunction`, which looks useless but isn't. And additionally, you'd have to provide trivial `none()` overrides of all the abstract predicates defined in `EscapeFunction`. This is all pretty awkward, and we can avoid it by distinguishing between `EscapeFunction` and `EscapeFunction::Range`.
```

Comment:
```suggestion
```

In the extra motivation I added above, I observe that you can't override the definitions of a member predicate contributed by extensions in a refining subclass that's implemented like this. Given that, this paragraph feels unnecessary. What do you think?

Reply:
I think if you're trying to wrap your head around this, I think this explanation is quite useful, so I would like to keep it.